### PR TITLE
Simplify construction of resolvers not attached to a project

### PR DESCRIPTION
**Pull request by jvandort**
This allows us to move DependencyMetaDataProvider and Module into dependency-management.

We further restrict detached resolvers from creating consumable configurations, to ensure they only represent resolvable root components and do not expose variants.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/StandaloneDomainObjectContext.java [hunk 1]
```diff
@@ -76,6 +76,49 @@ public String getDisplayName() {
...
+            }
+
+            @Override
+            public boolean isDetachedState() {
+                return true;
+            }
+        };
+    }
+
+    public static StandaloneDomainObjectContext forProjectBuildscript(ProjectInternal project) {
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/StandaloneDomainObjectContext.java:102]: Add javadoc

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/plugin/devel/variants/TargetJVMVersionOnPluginTooNewFailureDescriberIntegrationTest.groovy [hunk 2]
```diff
@@ -101,7 +101,7 @@ class TargetJVMVersionOnPluginTooNewFailureDescriberIntegrationTest extends Abst
         fails 'greet', "--stacktrace"
 
         then:
-        failure.assertHasErrorOutput("""> Could not resolve all artifacts for configuration ':classpath'.
+        failure.assertHasErrorOutput("""> Could not resolve all artifacts for configuration 'classpath'.
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/plugin/devel/variants/TargetJVMVersionOnPluginTooNewFailureDescriberIntegrationTest.groovy:104]: It should say the context somewhere. 

We want it to be clear which project buildscript had the problem if we're not adding a path now

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java [hunk 3]
```diff
@@ -171,18 +175,69 @@ public DefaultDependencyManagementServices(ServiceRegistry parent) {
...
+            fileCollectionFactory,
+            owner,
+            new AnonymousModule()
+        );
+
+        // We restrict this so that detached resolvers only represent adhoc root components that do not expose variants.
+        services.getConfigurationContainer().configureEach(configuration -> {
+            if (configuration.isCanBeConsumed()) {
+                throw new InvalidUserCodeException("Cannot create consumable configurations in detached resolvers");
+            }
```
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:194]: ❓ Will this catch every case of consumable configurations?  Can't a script create a non-consumable configuration and then register its own `configureEach` block (which will be run after this one) to make it consumable?

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/plugin/devel/variants/TargetJVMVersionOnPluginTooNewFailureDescriberIntegrationTest.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/plugin/devel/variants/TargetJVMVersionOnPluginTooNewFailureDescriberIntegrationTest.groovy:104]: For projects it is clearly stated.

For settings, standalone script, these have bad messages. 
However, removing the `:` from projects does not change that settings/init have bad messages.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java (hunk 3 above)
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java:194]: It is technically possible to bypass this, but `newDetachedResolver` is only used by internal code so this is more of a safeguard from ourselves. This should protect from most accidental creations of consumable configurations except those that really try to abuse this.

**Comment by jvandort** [comment]
@bot-gradle test this

**Pull request merged**